### PR TITLE
[JENKINS-25046] Drop cookie sessions when logging out

### PR DESCRIPTION
**Pull request by jsoref**
A user reported a `431 Header Fields Too Large`
The cause was lots of accumulated `JSESSION` records in the user's cookie.

As an admin, I suggested logging out, because, ideally, that should clear the user's
cookies.

Except it didn't do that.

Now it should...

See [JENKINS-25046](https://issues.jenkins-ci.org/browse/JENKINS-25046).

### Proposed changelog entries

* [JENKINS-25046] Drop sessions when logging out

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck

### CODE REVIEW

**Comment by daniel-beck** [comment]
How will this impact multiple Jenkinses running on the same host, different context paths?

**Comment by Wadeck** [comment]
> different context paths?

The path of the cookie should be checked, so that you remove only the related cookies instead of all.

**Comment by jsoref** [comment]
Technically, as long as they aren't sub mounts, I think the only cookies the browser would send would be the ones for the right Jenkins. But given that the information is available, I'll do the checking.

**Review by jvz — APPROVED** [review]
Oh this would be nice to see. I've been bit by this bug before in development.

**Comment by jvz** [comment]
However, some tests might be nice here. Wadeck has a good point about multiple Jenkinses (which may be more of an issue during development than in production depending on how you set up Jenkins).

**Comment by jsoref** [comment]
I started to try writing tests for this and hit some bumps

File: core/src/main/java/hudson/security/SecurityRealm.java
Comment (varyvol) [inline core/src/main/java/hudson/security/SecurityRealm.java]: Why the block?

Comment (jsoref) [reply core/src/main/java/hudson/security/SecurityRealm.java]: I used it to distinguish the two of users of `Cookie cookie`, but I guess that isn't really necessary.

**Review by jeffret-b — APPROVED** [review]
Looks good. I'd like to see the unnecessary block removed, as Evaristo mentioned. It would be great to get this cleaned up.

File: core/src/main/java/hudson/security/SecurityRealm.java
Comment (jsoref) [reply core/src/main/java/hudson/security/SecurityRealm.java]: I take that back, it is necessary. Java won't allow me to not type my cookie in the for loop, and it won't let me declare the same object both in the function scope and in the for loop scope.

That alternative is that I define two differently named cookie variables. If you have a preference between the two, I'm open to suggestions. Personally, I think using scoped variables is better than naming one `cookie2` or `c` or whatever.

Comment (jeffret-b) [reply core/src/main/java/hudson/security/SecurityRealm.java]: I would probably name the one in the for loop on line 303 something like "currentCookie". Or maybe pull one or both parts of the behavior out into methods that describe what they're doing. That's probably my favorite.

But, this is getting to the level of personal preference and I really don't care which one you go with.

**Comment by jsoref** [comment]
both blobs needed access to the same three things (req, rsp, contextPath), so now instead of having an "unnecessary" block, I've gone with two functions.

**Comment by jeffret-b** [comment]
I like that approach. Now it's much easier to tell what the different parts are doing.

File: core/src/main/java/hudson/security/SecurityRealm.java
Comment (Wadeck) [inline core/src/main/java/hudson/security/SecurityRealm.java]: 1) From my PoV the `continue` keyword should be avoided to prevent troubles.
2) The `if` should use `{}`
3) Consistency is really important to help readability of code

Proposal:
```
for (Cookie cookie : req.getCookies()) {
    if (cookie.getName().startsWith("JSESSIONID.")) {
        String cookieContextPath = cookie.getPath();
        boolean pathMatches;
        if (cookieContextPath == null) {
            pathMatches = contextPath.equals("/");
        } else {
            pathMatches = cookieContextPath.equals(contextPath);
        }
        if (pathMatches) {
            // one reason users log out is to clear their session(s)
            // so tell the browser to drop all old sessions
            cookie.setMaxAge(0);
            cookie.setValue("");
            rsp.addCookie(cookie);
        }
    }
}
```

**Review by Wadeck — CHANGES_REQUESTED** [review]
After re-reading the documentation around the JSESSIONID, I found back: https://github.com/jenkinsci/extras-executable-war/#jetty-session-ids.

Due to that, I think your solution will not work at 100%. I think @oleg-nenashev (for the session customization) or even @olamy (for the way Jetty is using the cli options) could help here. If you do not take into consideration the options of the application server, the fix is not complete.

In addition, did you try to have multiple Jenkins running on the same host as explained in the doc?

Note: IIRC the session cookie is not managed by Jenkins but by the application server directly and thus, we cannot guarantee anything around the use of the Path. If I am right, the check on the path is not sufficient to ensure we are not killing the session of the users from all the running Jenkins under that host (think of https://internal/jenkins-team-a, https://internal/jenkins-team-b)

**Comment by jsoref** [comment]
I wasn't claiming this was a 100% fix. Putting something into the log out party is a desperation fix. I was also looking to put something into the log in path later if I could find it. But if there's a place closer to where these tokens are managed where this code would be more appropriate, I'm happy to move it.

**Comment by jsoref** [comment]
So, I've looked and both my jenkins and ci.jenkins.io appear to be issuing session cookies, not time limited cookies.

https://github.com/jenkinsci/winstone/blob/04bcf4e65a19ceacb93df9dcec8fefe3b638a95d/src/main/java/winstone/HostConfiguration.java#L187

Afaict, this has been the case for a *long* time.... But, I could be missing something.

**Comment by jsoref** [comment]
tbh, I really don't think I or anyone else should care about the hazard of `log out` killing other session cookies. The user is actively driving this click, which means they're more or less present. Worst case, multiple jenkins that the user is accessing get logged out, and when they try to access another, they have to log in again. Jenkins isn't mission critical. Having to log in again to continue performing an action isn't the end of the world.

At this point, the burden of proof and work should be on people who are preventing users from actually using their jenkins, not me to find a perfect fix.

And really, at this point, if you want to suggest a change to my PR (or anyone else's), please use github's suggested change feature. Otherwise, you're really just wasting everyone's time.

My patience is wearing very thin. I just took a week-long break. My next one will be for longer.

File: core/src/main/java/hudson/security/SecurityRealm.java
Comment (michaelneale) [reply core/src/main/java/hudson/security/SecurityRealm.java]: app servers and multiple paths in 2019 seem fairly edge cases (certainly the former) for the pain caused by this. If the worst case is a user session log out, not that bad?

**Comment by oleg-nenashev** [comment]
needs re-review from @Wadeck

File: core/src/main/java/hudson/security/SecurityRealm.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/security/SecurityRealm.java]: Note that the Cookie ID might be different : https://github.com/jenkinsci/extras-executable-war/blob/6558df699d1366b18d045d2ffda3e970df377873/src/main/java/Main.java#L79-L97

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Looks good to me overall, but the hardcoded cookie prefix will not work in all cases. https://github.com/jenkinsci/extras-executable-war/blob/6558df699d1366b18d045d2ffda3e970df377873/src/main/java/Main.java#L79-L97

**Comment by jsoref** [comment]
So, I'm trying to switch from using ATH to having the test run locally. In doing so, the default test land shows that while jenkins lives as `/jenkins`, the cookie jenkins issues has no path.

The right behavior is to trust the browser. See [Scope of cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Scope_of_cookies).

Consider a site with:
1. `/jenkins-a`
1. `/jenkins-b`
The sites should each be configured to provide their own paths of `/jenkins-a` or `/jenkins-b`. When the browser visits `/jenkins-a`, it will not send a cookie for `/jenkins-b` even if it has such a cookie.

If a site has:
1. `/` (a jenkins) and
1. `/jenkins` (a different jenkins)
Then when the browser visits `/jenkins` it will only send cookies for that instance, which is correct. When the browser visits `/`, it must send all the cookies, and this is just a misconfigured system -- the fact that users get logged out from the second jenkins would be a good thing to help their administrators fix their configuration.

**Comment by oleg-nenashev** [comment]
Would it be possible to avoid force-pushing to the branches while they are under review? It complicates the review process, because I cannot see diffs since the previous review. We can always squash-merge the changes once the PR is reviewed.

File: core/src/main/java/hudson/security/SecurityRealm.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/security/SecurityRealm.java]: By default Jetty would set just `JSESSIONID` as a cookie name. So maybe dot should be removed from here

Comment (jsoref) [reply core/src/main/java/hudson/security/SecurityRealm.java]: So, thinking about this, I'm not sure it's the right behavior.

My goal is to delete all the stale cookies.

1. Let's say you have a server which was configured with cookieName with the jenkins default (`JSESSIONID.`+_random_)
2. User visits and server and over time (as it's restarted or whatever) collects cookies.
3. Someone turns off the feature (and restarts jenkins).
4. Now you have a whole bunch of `JSESSIONID.`+_random_ cookies which will never go away.
5. User is stuck and tries to log out -- this feature will not help them if we only look at the current configuration.

Remember, the goal is not to delete the current jenkins session cookie, jenkins is already managing (and **using**) that cookie. The goal is to delete all the other ones.

Comment (jsoref) [reply core/src/main/java/hudson/security/SecurityRealm.java]: For my reference, the correct code to try to emulate the underlying behavior would have been:
```
        if (!DISABLE_CUSTOM_JSESSIONID_COOKIE_NAME) {
            if (JSESSIONID_COOKIE_NAME != null) {
                cookieName = JSESSIONID_COOKIE_NAME;
            } else {
                cookieName = "JSESSIONID.";
            }
        } else {
            cookieName = "JSESSIONID";
        }
```
Or to just use the `Field` setting code to look up the value, but, as noted above, that's pointless, as it doesn't accomplish the actual goal here.

**Comment by jsoref** [comment]
@oleg-nenashev: I'm done working on this PR. Please review and squash-merge.

---
If people have objections, at this point, I'm really not interested in them. While there has been some change between my original code and this final version, none of it functional:
* refactoring two code blocks
* replacing a `continue` with a block
* adding copious lines of comments
* adding an entirely useless test (because it isn't sufficient and at the end of the day, we really need the acceptance test changes to have any confidence that it's working properly)

https://github.com/jenkinsci/jenkins/compare/3b94cc8fe1988ad317b9cb2634664e612a75f5da..c436b9360ae7c106d9def853ed4cda16b08143da#diff-134

**Comment by oleg-nenashev** [comment]
@jsoref with all due respect, I consider comments like "If people have objections, at this point, I'm really not interested in them" as borderline ones. Code review in Jenkins is a public process open to contributors with all level of experience, and code reviews also serve purpose of sharing the experience about the codebase. Sometime it leads to frictions and not-so-helpful comments, but in the Jenkins project we expect all contributors to communicate in a respectful way. Such aggressive comments do not help

**Comment by jsoref** [comment]
@oleg-nenashev : I'm sorry about that. I understand we're all volunteers. But the iterations here over the course of a month have not been helpful to this PR. And I'm wearing out. -- I'm about to switch focal interests to another project (or two or three).

Reviews being open to all levels of experience is a lofty goal. But sometimes it'd be nice if people looked into how a thing worked before asking questions/for changes. -- That especially applies to the cookie path bits in this case.

**Comment by jsoref** [comment]
@Wadeck : can you please re-review this?

File: test/src/test/java/hudson/security/SecurityRealmTest.java
Comment (Wadeck) [inline test/src/test/java/hudson/security/SecurityRealmTest.java]: unused variable
```suggestion
```

(⚠️ first time I tried the suggestion feature to remove a line)

**Review by Wadeck — APPROVED** [review]
Not tested manually but due to the discussion and the code, it seems fine.

Note for admin that are coming here from the future, having troubles with the sessions from other Jenkins instance to be removed, you can follow the advice from [here](https://github.com/jenkinsci/extras-executable-war/#jetty-session-ids) and set different `sessionIdCookieName` per instance.

@jsoref, I was unable to remember what I reviewed here exactly (45 days ago?) and with all the force pushes that has not helped at all. If you have the choice, avoid force pushing, that reduces the tooling capability.

Thank you for the test if it was not present in the initial proposal.

**Comment by jsoref** [comment]
My original PR was a work in progress followed by an acceptance test as I couldn't figure out how to do in repository testing if this. At the end, I've concluded that the acceptance tests are much more valuable than the in repository tests, as they actually test the right Jenkins configurations (the random ID stuff).

**Review by oleg-nenashev — COMMENTED** [review]
I am still concerned about not removing the custom session ID cookies if the instance has a custom configuration. There is a long explanation in https://github.com/jenkinsci/jenkins/pull/3939/commits/bf65e1b4b40356b6a19a461ae2396a76cc3c08e0 where the suggested approach was removed, but IMHO "different incarnation" is not enough. The session management code was in the codebase for a while.

If there is a consensus to go forward with the current approach, I would expect to at least have FINE(or less) System logging which would help admins to troubleshoot cookie management issues.

> We don't care about JSESSIONID cookies outside our path because it's the browser's responsibility not to send them to us in the first place.

Removing stale cookies is also the browser's responsibility...

**Comment by jsoref** [comment]
The problem with removing stale cookies is that they're marked as session cookies and the web (and especially users) evolved with an expectation not to expire such cookies.  https://stackoverflow.com/questions/10617954/chrome-doesnt-delete-session-cookies

**Comment by jvz** [comment]
Agreed with Josh. Besides, this problem comes up every once in a while particularly for development. Whenever I use `mvn hpi:run` or `mvn -f war hudson-dev:run`, that defaults to port 8080, so my browser builds up a fairly large cookie for `localhost:8080/jenkins` or just `localhost:8080` over time. The first time I got a "response entity too large" HTTP error, I thought it was a bug in Jenkins. 🤷‍♂

**Review by timja — APPROVED** [review]
This solves the majority of cases. I've been bitten by this before in local development...

**Comment by batmat** [comment]
Given approvals and especially last comments, this seems like it is good enough, and has four approvals already. Users seem to still be hit by this, and my understanding is this would improve the life of the (vast?) majority of them.

Are we actually aware of some customization of Jetty around here?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
>If there is a consensus to go forward with the current approach, I would expect to at least have FINE(or less) System logging which would help admins to troubleshoot cookie management issues

File: core/src/main/java/hudson/security/SecurityRealm.java
Comment (timja) [inline core/src/main/java/hudson/security/SecurityRealm.java]: This may satisfy @oleg-nenashev 's concerns
```suggestion
                LOGGER.log(Level.FINE, "Removing cookie {0} during logout", cookie.getName());
                // one reason users log out is to clear their session(s)
```

also needs this import
```
import java.util.logging.Level;
```

**Comment by jeffret-b** [comment]
@jsoref, when you merged in @timja's logging suggestion, I think you missed their other comment, 
> also needs this import
> 
> ```
> import java.util.logging.Level;
> ```

This currently has a compilation failure around that.

**Comment by jeffret-b** [comment]
I'm still a +1 on this change. It's not perfect but it should make a noticeable improvement for some users in some cases. The problems with this proposal are mostly along the lines of it missing cleanup on a few things sometimes. At least it will help in some cases, even if we can't get it perfect. The cases where it might go too far, with the user logging out, seem reasonably justified.

**Comment by jsoref** [comment]
@jeffret-b : I didn't miss it, I used that to get his commit info, and then I made the commit and ran into interruptions which prevented me from actually force pushing.

**Comment by jeffret-b** [comment]
That's fine. I meant that it didn't get included in the changes at the time I was taking a look. Wanted to make sure you noticed it.

**Review by oleg-nenashev — APPROVED** [review]
Thanks for adding logging. the commit history got messed up after the force push. But the code itself looks good to me. If there is an intention to backport it, please rebase the changes.

**Review by oleg-nenashev — APPROVED** [review]
Thanks for adding logging. the commit history got messed up after the force push. But the code itself looks good to me. If there is an intention to backport it, please rebase the changes.

**Comment by batmat** [comment]
image

Given the current status now: 5 approvals, and no change request anymore, I think we can move forward.

Let's merge this tomorrow or so if nobody objects in the meantime.

Thanks a lot everyone, especially @jsoref !

**Comment by oleg-nenashev** [comment]
> If there is an intention to backport it, please rebase the changes.

Just to highlight it. But I agree with merging as is as well

**Comment by daniel-beck** [comment]
@jsoref Please be more careful when rebasing your PRs. This PR includes unrelated rebased commits from 2.183 that are included here: https://github.com/jenkinsci/jenkins/pull/3939/commits

It should never be really necessary to rebase in the first place, just merge with master. Thanks!

If you continue to do this anyway, please also make sure to not rebase to the weekly "release" commit, instead choose one with a snapshot version.

Some of our tooling behaves weirdly of you do any of the above, creating additional work.

**Comment by michaelneale** [comment]
kudos @jsoref and all for making this happen - a long standing point of confusion and annoyance. 

More cookies are good to eat, not in the browser.

**Pull request merged**
